### ISSUE TITLE
Is option std_cxx_17 required ?

### ISSUE THREAD

**Issue by stephan57160** [corrosion-rs/corrosion#183]
I failed to rebuild `corrosion` on CentOS-7.x-based system.
One of the reasons is the rather old compiler versions on this system.
Please, don't laugh, but I use this old stuff for many years and other projects.

Anyway, in this context `corrosion` build process fails with
```
CMake Error at test/cpp2rust/CMakeLists.txt:5 (target_compile_features):
  target_compile_features The compiler feature "cxx_std_17" is not known to
  CXX compiler

  "GNU"

  version 4.8.5.
```
...

In `test/cpp2rust/CMakeLists.txt`, I changed `std_cxx_17` to `std_cxx_14`.
With the help of some minor other things, I managed to have compilation till the end.

Before I raise a PR, I'd juste like to know if there is any reason for this CMAKE option ?
Is it a requirement from RUST, or from anything else ?
Would this break anything, if removed ?

**Comment by ogoffart** [comment]
This is just needed for this one test. It uses string_view which is C++17.

You shouldn't need C++17 to use corrosion though?
How are you using Corrosion? Are you explicitly turning on the tests?

**Comment by stephan57160** [comment]
Well ... I plan to use it, yes, but still in evaluation phase, with some other things.

I came to this, with the "build" process given in the README, nothing more, nothing less.

By the way, the minor other things I mentioned, are related to <string_view> in the 3 test cases :-)
Had to drop its use in `lib.cpp`, `lib2.cpp` and `lib3.cpp` ...

If really not required for `corrosion` to work, I may send a PR.
Just tell me, else, I keep the patch for myself, at least, for now.

**Comment by ogoffart** [comment]
I think maybe the problem is that the tests are build by default if corrosion is built from the root directory.  Maybe `CORROSION_BUILD_TESTS` should always default to false, and we should document to enable the tests for developer wanting to hack on corrosion.

In practice though, if you simply want to use corrosion, you probably will want to use `FetchContent` or bring corrosion in a sub-directory.   In that case, the tests are not build by default.


Anyway, regarding your direct question, I think it'd be alright to port the tests away from C++17.

**Comment by jschwe** [comment]
Another way would be to use CMake to test if 
a) There is a C++ compiler on the system, and
b) If the C++ compiler is recent enough
and skip the test in question if that is not the case.
But I'm also not opposed to changing the test code to something that is compatible with older C++ versions. The only requirement is that the test code should require the C++ standard library (so that linking the test fails if e.g. the C-linker is selected by corrosion).

> In practice though, if you simply want to use corrosion, you probably will want to use FetchContent or bring corrosion in a sub-directory.

I just want to add, that if your CMake version is older than 3.19, you probably still want to install Corrosion, since otherwise the configuration step will be more expensive, due to building a Rust tool that Corrosion needs for CMake < 3.19.

**Comment by stephan57160** [comment]
Unfortunatelly, my CMake is 3.18 :-)
I could rebuild a 3.19, but this would break some other things, for us.

> In practice though, if you simply want to use corrosion, you probably will want to use FetchContent or bring corrosion in a sub-directory.

FetchContent not avail.

> you probably still want to install Corrosion

This is the case and done, now. 
I need to test this setup, now :-)

Regarding Corrosion and test cases, in general, my reasonning is that any test should test what is actually done or required by the tool/application/library to be tested.

To avoid any risk of broken stuff, I only replaced `std_cxx_17` by `std_cxx_14`, and removed the use of <string_view>.
I did not try to go for an older version ... Then, we still have some C++ compiler check. I suppose that `std_cxx_14` will
make sure some (maybe old) version of C++ Standard Library is installed.
